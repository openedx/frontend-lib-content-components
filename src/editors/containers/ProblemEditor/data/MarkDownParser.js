/*
This is the direct motivation code taken from:
https://github.com/open-craft/capa-visual-editor/

Changes have been made to adapt the parser with respect to our use case.
*/

import { ProblemTypeKeys } from "../../../data/constants/problem";

export const indexToLetterMap = [...Array(26)].map((val, i) => String.fromCharCode(i + 65));

function getHints(markdown) {
  if (!markdown || !markdown.trim()) {
    return [];
  }
  const hints = [];

  function getHint(row) {
    let hint = '';
    if (row.startsWith('||')) {
      const hintStart = row.indexOf('||');
      const hintEnd = row.slice(hintStart + 2).indexOf('||');
      hint = row.slice(hintStart + 2, hintEnd + 2);
    }
    return hint;
  }
  for (const d in markdown.split('\n')) {
    const row = markdown.split('\n')[d];
    const hint = getHint(row);
    if (hint.length) {
      hints.push({
        id: Number(d),
        value: hint,
      });
    }
  }
  return hints;
}

function getShortAnswerOptions(markdown) {
  // TODO: Include tolerence for number it fails for now.
  let problemType = ProblemTypeKeys.TEXTINPUT;
    if ((!markdown || !markdown.trim())) {
      return {
        shortAnswersList: [],
        problemType,
      };
    }
    const markdownListData = markdown.split('\n');
    const shortAnswersList = [];
    let gotAnswer = false;
    for (const i in markdownListData) {
      const row = markdownListData[i].trim();
      const feedbackStart = row.indexOf('{{');
      const feedbackEnd = row.indexOf('}}');
      let feedback = '';
      let answer = '';
      let correct = true;
      if (row.startsWith('=') && !gotAnswer) {
        gotAnswer = true;
        if (feedbackStart > -1) {
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
          answer = row.slice(1, feedbackStart).trim();
        } else {
          answer = row.slice(1).trim();
        }
      } else if (row.startsWith('=') && gotAnswer) {
        break;
      } else if (gotAnswer && row.startsWith('or=')) { // the next correct answer
        if (feedbackStart > -1) {
          answer = row.slice(3, feedbackStart).trim();
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
        } else {
          answer = row.slice(3).trim();
        }
      } else if (gotAnswer && row.startsWith('not=')) { // the incorrect answer
        if (feedbackStart > -1) {
          answer = row.slice(4, feedbackStart).trim();
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
        } else {
          answer = row.slice(4).trim();
        }
        correct = false;
      }
      if (answer) {
        problemType = isFinite(answer) ? ProblemTypeKeys.NUMERIC : ProblemTypeKeys.TEXTINPUT;
        shortAnswersList.push({
          id: indexToLetterMap[shortAnswersList.length],
          title: answer,
          feedback,
          correct,
        });
      }
    }
    return {
      shortAnswersList,
      problemType,
    };
}

function getMultipleChoiceOptions(markdown) {
  const multipleChoiceOptions = [];
  const data = {
    multiSelectAnswersList: multipleChoiceOptions,
    problemType: ProblemTypeKeys.MULTISELECT,
    groupFeedbackList: [],
  };
  if (!markdown || !markdown.trim()) {
    return data;
  }
  const markdownListData = markdown.split('\n');
  let feedback = '';
  for (const d in markdownListData) {
    const row = markdownListData[d];
    // multiple choices
    if (row.startsWith('[ ]') || row.startsWith('[x]')) {
      // title
      let title = row.slice(row.indexOf(']') + 1).trim();

      const feedbacksStart = row.indexOf('{{');
      const selectedFeedbackStart = row.indexOf(' selected:') || row.indexOf('{selected:'); // two ways to define selected feedback
      const selectedFeedbackEnd = row.slice(selectedFeedbackStart + 1).indexOf('}') + selectedFeedbackStart;
      const unselectedFeedbackStart = row.indexOf('unselected:');
      const unselectedFeedbackEnd = row.slice(unselectedFeedbackStart + 1).indexOf('}}') + unselectedFeedbackStart;

      let selectedFeedback = '';
      let unselectedFeedback = '';
      let titleChanged = false;
      if (feedbacksStart !== -1 && selectedFeedbackStart !== -1) {
        selectedFeedback = row.slice(selectedFeedbackStart + 10, selectedFeedbackEnd);
        title = row.slice(row.indexOf(']') + 1, feedbacksStart).trim();
        titleChanged = true;
      }
      if (feedbacksStart !== -1 && unselectedFeedbackStart !== -1) {
        unselectedFeedback = row.slice(unselectedFeedbackStart + 12, unselectedFeedbackEnd + 1);
        if (!titleChanged) {
          title = row.slice(row.indexOf(']') + 1, feedbacksStart).trim();
        }
      }
      multipleChoiceOptions.push({
        id: indexToLetterMap[multipleChoiceOptions.length],
        title,
        correct: !row.startsWith('[ ]'),
        selectedFeedback,
        unselectedFeedback,
      });
    }
    // group Feedback
    if (multipleChoiceOptions.length) {
      if (row.startsWith('{{')) {
        const groupFeedbackEnd = row.indexOf('}}');
        const goupStart = row.indexOf('((');
        const groupEnd = row.indexOf('))');
        const group = row.slice(goupStart + 2, groupEnd).split(' ');
        const groupChoices = [];
        for (const i in group) {
          const letter = group[i];
          if (multipleChoiceOptions[indexToLetterMap.indexOf(letter)]) {
            groupChoices.push(multipleChoiceOptions[indexToLetterMap.indexOf(letter)].id);
          }
        }
        feedback = row.slice(groupEnd + 2, groupFeedbackEnd).trim();

        if (groupChoices.length && feedback) {
          data.groupFeedbackList.push({
            id: groupChoices.length,
            answers: groupChoices,
            feedback,
          });
        }
      }
    }
  }
  return data;
}

function getSingleChoiceOptions(markdown) {
  // TODO: Build support for selected and unselected feedback
  if (!markdown || !markdown.trim()) {
    return {
      singleSelectAnswersList: [],
      problemType: ProblemTypeKeys.DROPDOWN,
    };
  }

  const singleChoiceOptions = [];
  let dropDownMode = false;
  const markdownListData = markdown.split('\n');
  for (const d in markdownListData) {
    const row = markdownListData[d].trim();
    if (row.startsWith('[[')) {
      dropDownMode = true;
      continue;
    }
    if (row.startsWith(']]')) {
      break;
    }
    if (dropDownMode) {
      if (row.startsWith('(')) {
        let feedback = '';
        let title = row.slice(row.indexOf('(') + 1, row.indexOf(')')).trim();
        const feedbackStart = row.indexOf('{{');
        const feedbackEnd = row.indexOf('}}');
        if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
          feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
          title = title.slice(0, feedbackStart).trim();
        }
        const correct = true;
        singleChoiceOptions.push({
          id: indexToLetterMap[singleChoiceOptions.length],
          title,
          correct,
          feedback,
        });
      } else {
        let feedback = '';
        let title = '';
        // let title = row.slice(row.indexOf(')') + 1).trim();

        const feedbackStart = row.indexOf('{{');
        const feedbackEnd = row.indexOf('}}');
        if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
          feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
          title = row.slice(0, feedbackStart).trim();
        } else {
          title = row.trim();
        }
        const correct = false;
        singleChoiceOptions.push({
          id: indexToLetterMap[singleChoiceOptions.length],
          title,
          correct,
          feedback,
        });
      }
    } else if (row.startsWith('( )') || row.startsWith('(x)')) {
      let feedback = '';
      let title = row.slice(row.indexOf(')') + 1).trim();
      const feedbackStart = row.indexOf('{{');
      const feedbackEnd = row.indexOf('}}');
      if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
        feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
        title = title.slice(0, feedbackStart - 4).trim();
      }
      const correct = !!row.startsWith('(x)');
      singleChoiceOptions.push({
        id: indexToLetterMap[singleChoiceOptions.length],
        title,
        correct,
        feedback,
      });
    }
  }

  return {
    singleSelectAnswersList: singleChoiceOptions,
    problemType: dropDownMode ? ProblemTypeKeys.DROPDOWN : ProblemTypeKeys.SINGLESELECT,
  };
}

function getEditorData(markdown) {
  // TODO: Prase heading in this function.
  if (!markdown || !markdown.trim()) {
    return '';
  }
  let description = '';
  const markdownData = markdown.trim().split('\n');
  for (const i in markdownData) {
    const row = markdownData[i];
    if (['{', '(', '[', '=', '|'].indexOf(row.trim()[0]) === -1) {
      const questionTextStart = row.indexOf('>>');
      const questionTextEnd = row.indexOf('<<');
      let questionText = row;
      if (questionTextStart !== -1 && questionTextEnd !== -1 && (questionTextEnd > questionTextStart)) {
        questionText = row.replace('>>', '<p>').replace('<<', '</p>');
      }
      description += `${questionText}\n`;
    } else {
      break;
    }
  }
  return description;
}

function parseMarkdown(markdown) {
  if (!markdown || !markdown.trim()) {
    return {};
  }
  let answers = {};
  let problemType = '';
  const question = getEditorData(markdown);
  const hints = getHints(markdown);
  let data = {question, settings: {hints}};
  let parsedData = getSingleChoiceOptions(markdown);
  if (parsedData.singleSelectAnswersList.length !== 0){
    answers = parsedData.singleSelectAnswersList;
    problemType = parsedData.problemType;
    data = {...data, answers, problemType};
  }
  parsedData = getMultipleChoiceOptions(markdown);
  if (parsedData.multiSelectAnswersList.length !== 0){
    answers = parsedData.multiSelectAnswersList;
    problemType = parsedData.problemType;
    const groupFeedbackList = parsedData.groupFeedbackList;
    data = {...data, answers, problemType, groupFeedbackList};
  }
  parsedData = getShortAnswerOptions(markdown);
  if (parsedData.shortAnswersList.length !== 0){
    answers = parsedData.shortAnswersList;
    problemType = parsedData.problemType;
    data = {...data, answers, problemType};
  }
  return data;
}

function getScorringSettings() {
  return {
    selectedAttemptsOption: window.LXCData.max_attempts,
    selectedPointOption: window.LXCData.weight,
  };
}

export {
  getMultipleChoiceOptions,
  getSingleChoiceOptions,
  getEditorData,
  getHints,
  getScorringSettings,
  getShortAnswerOptions,
  parseMarkdown,
};
