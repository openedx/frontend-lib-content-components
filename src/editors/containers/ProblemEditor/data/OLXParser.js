// Parse OLX to JavaScript objects.
/* eslint no-eval: 0 */

import { XMLParser, XMLBuilder } from 'fast-xml-parser';
import _ from 'lodash-es';
import { ProblemTypeKeys } from '../../../data/constants/problem';

export const indexToLetterMap = [...Array(26)].map((val, i) => String.fromCharCode(i + 65));

export const nonQuestionKeys = [
  '@_answer',
  '@_type',
  'additional_answer',
  'checkboxgroup',
  'choicegroup',
  'choiceresponse',
  'correcthint',
  'demandhint',
  'formulaequationinput',
  'multiplechoiceresponse',
  'numericalresponse',
  'optioninput',
  'optionresponse',
  'responseparam',
  'solution',
  'stringequalhint',
  'stringresponse',
  'textline',
];

export class OLXParser {
  constructor(olxString) {
    this.problem = {};
    const options = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
      // preserveOrder: true
    };
    const parser = new XMLParser(options);
    this.parsedOLX = parser.parse(olxString);
    if (_.has(this.parsedOLX, 'problem')) {
      this.problem = this.parsedOLX.problem;
    }
  }

  parseMultipleChoiceAnswers(problemType, widgetName, option) {
    const answers = [];
    let data = {};
    const widget = _.get(this.problem, `${problemType}.${widgetName}`);
    const choice = _.get(widget, option);
    if (_.isArray(choice)) {
      choice.forEach((element, index) => {
        const title = element['#text'];
        const correct = eval(element['@_correct'].toLowerCase());
        const id = indexToLetterMap[index];
        const feedback = this.getAnswerFeedback(element, `${option}hint`);
        answers.push(
          {
            id,
            title,
            correct,
            ...feedback,
          },
        );
      });
    } else {
      const feedback = this.getAnswerFeedback(choice, `${option}hint`);
      answers.push({
        title: choice['#text'],
        correct: eval(choice['@_correct'].toLowerCase()),
        id: indexToLetterMap[answers.length],
        ...feedback,
      });
    }
    data = { answers };
    const groupFeedbackList = this.getGroupedFeedback(widget);
    if (groupFeedbackList.length) {
      data = {
        ...data,
        groupFeedbackList,
      };
    }
    return data;
  }

  getAnswerFeedback(choice, hintKey) {
    let feedback = {};
    let feedbackKeys = 'feedback';
    if (_.has(choice, hintKey)) {
      const answerFeedback = choice[hintKey];
      if (_.isArray(answerFeedback)) {
        answerFeedback.forEach((element) => {
          if (_.has(element, '@_selected')) {
            feedbackKeys = eval(element['@_selected'].toLowerCase()) ? 'selectedFeedback' : 'unselectedFeedback';
          }
          feedback = {
            ...feedback,
            [feedbackKeys]: element['#text'],
          };
        });
      } else {
        if (_.has(answerFeedback, '@_selected')) {
          feedbackKeys = eval(answerFeedback['@_selected'].toLowerCase()) ? 'selectedFeedback' : 'unselectedFeedback';
        }
        feedback = {
          [feedbackKeys]: answerFeedback['#text'],
        };
      }
    }
    return feedback;
  }

  getGroupedFeedback(choices) {
    const groupFeedback = [];
    if (_.has(choices, 'compoundhint')) {
      const groupFeedbackArray = choices.compoundhint;
      if (_.isArray(groupFeedbackArray)) {
        groupFeedbackArray.forEach((element) => {
          groupFeedback.push({
            id: groupFeedback.length,
            answers: element['@_value'].split(' '),
            feedback: element['#text'],
          });
        });
      } else {
        groupFeedback.push({
          id: groupFeedback.length,
          answers: groupFeedbackArray['@_value'].split(' '),
          feedback: groupFeedbackArray['#text'],
        });
      }
    }
    return groupFeedback;
  }

  parseStringResponse() {
    const { stringresponse } = this.problem;
    const answers = [];
    let answerFeedback = '';
    let additionalStringAttributes = {};
    let data = {};
    const feedback = this.getFeedback(stringresponse);
    answers.push({
      id: indexToLetterMap[answers.length],
      title: stringresponse['@_answer'],
      correct: true,
      feedback,
    });

    // Parsing additional_answer for string response.
    const additionalAnswer = _.get(stringresponse, 'additional_answer', []);
    if (_.isArray(additionalAnswer)) {
      additionalAnswer.forEach((newAnswer) => {
        answerFeedback = this.getFeedback(newAnswer);
        answers.push({
          id: indexToLetterMap[answers.length],
          title: newAnswer['@_answer'],
          correct: true,
          feedback: answerFeedback,
        });
      });
    } else {
      answerFeedback = this.getFeedback(additionalAnswer);
      answers.push({
        id: indexToLetterMap[answers.length],
        title: additionalAnswer['@_answer'],
        correct: true,
        feedback: answerFeedback,
      });
    }

    // Parsing stringequalhint for string response.
    const stringEqualHint = _.get(stringresponse, 'stringequalhint', []);
    if (_.isArray(stringEqualHint)) {
      stringEqualHint.forEach((newAnswer) => {
        answers.push({
          id: indexToLetterMap[answers.length],
          title: newAnswer['@_answer'],
          correct: false,
          feedback: newAnswer['#text'],
        });
      });
    } else {
      answers.push({
        id: indexToLetterMap[answers.length],
        title: stringEqualHint['@_answer'],
        correct: false,
        feedback: stringEqualHint['#text'],
      });
    }

    // TODO: Support multiple types.
    additionalStringAttributes = {
      type: _.get(stringresponse, '@_type'),
      textline: {
        size: _.get(stringresponse, 'textline.@_size'),
      },
    };

    data = {
      answers,
      additionalStringAttributes,
    };

    return data;
  }

  parseNumericResponse() {
    const { numericalresponse } = this.problem;
    let answers = [];
    let subAnswers = [];
    let data = {};
    // TODO: Find a way to add answers using additional_answers v/s numericalresponse
    if (_.isArray(numericalresponse)) {
      numericalresponse.forEach((numericalAnswer) => {
        subAnswers = this.parseNumericResponseObject(numericalAnswer, answers.length);
        answers = _.concat(answers, subAnswers);
      });
    } else {
      subAnswers = this.parseNumericResponseObject(numericalresponse, answers.length);
      answers = _.concat(answers, subAnswers);
    }

    data = {
      answers,
    };

    return data;
  }

  parseNumericResponseObject(numericalresponse, answerOffset) {
    let answerFeedback = '';
    const answers = [];
    let responseParam = {};
    // TODO: UI needs to be added to support adding tolerence in numeric response.
    const feedback = this.getFeedback(numericalresponse);
    if (_.has(numericalresponse, 'responseparam')) {
      const type = _.get(numericalresponse, 'responseparam.@_type');
      const defaultValue = _.get(numericalresponse, 'responseparam.@_default');
      responseParam = {
        [type]: defaultValue,
      };
    }
    answers.push({
      id: indexToLetterMap[answers.length + answerOffset],
      title: numericalresponse['@_answer'],
      correct: true,
      feedback,
      ...responseParam,
    });

    // Parsing additional_answer for numerical response.
    const additionalAnswer = _.get(numericalresponse, 'additional_answer', []);
    if (_.isArray(additionalAnswer)) {
      additionalAnswer.forEach((newAnswer) => {
        answerFeedback = this.getFeedback(newAnswer);
        answers.push({
          id: indexToLetterMap[answers.length + answerOffset],
          title: newAnswer['@_answer'],
          correct: true,
          feedback: answerFeedback,
        });
      });
    } else {
      answerFeedback = this.getFeedback(additionalAnswer);
      answers.push({
        id: indexToLetterMap[answers.length + answerOffset],
        title: additionalAnswer['@_answer'],
        correct: true,
        feedback: answerFeedback,
      });
    }
    return answers;
  }

  parseQuestions(problemType) {
    const builder = new XMLBuilder();
    const problemObject = _.get(this.problem, problemType);
    let questionObject = {};
    /* TODO: How do we uniquely identify the label and description?
      In order to parse label and description, there should be two states
      and settings should be introduced to edit the label and description.
      In turn editing the settings update the state and then it can be added to
      the parsed OLX.
    */
    const tagMap = {
      description: 'em',
    };

    /* Only numerical response has different ways to generate OLX, test with
      numericInputWithFeedbackAndHintsOLXException and numericInputWithFeedbackAndHintsOLX
      shows the different ways the olx can be generated.
    */
    if (_.isArray(problemObject)) {
      questionObject = _.omitBy(problemObject[0], (value, key) => _.includes(nonQuestionKeys, key));
    } else {
      questionObject = _.omitBy(problemObject, (value, key) => _.includes(nonQuestionKeys, key));
    }
    // Check if problem tag itself will have question and descriptions.
    if (_.isEmpty(questionObject)) {
      questionObject = _.omitBy(this.problem, (value, key) => _.includes(nonQuestionKeys, key));
    }
    const serializedQuestion = _.mapKeys(questionObject, (value, key) => _.get(tagMap, key, key));

    const questionString = builder.build(serializedQuestion);
    return questionString;
  }

  getHints() {
    const hintsObject = [];
    if (_.has(this.problem, 'demandhint.hint')) {
      const hint = _.get(this.problem, 'demandhint.hint');
      if (_.isArray(hint)) {
        hint.forEach(element => {
          hintsObject.push({
            id: hintsObject.length,
            value: element['#text'],
          });
        });
      } else {
        hintsObject.push({
          id: hintsObject.length,
          value: hint['#text'],
        });
      }
    }
    return hintsObject;
  }

  #extractTextAndChildren(node) {
    const children = [];
    let text = null;

    if (_.isArray(node)) {
      children.push(...node);
    } else if (_.isPlainObject(node)) {
      text = _.get(node, '#text');
      const nodeWithoutText = _.omit(node, '#text');
      children.push(...Object.values(nodeWithoutText));
    }

    return { text, children };
  }

  getSolutionExplanation() {
    if (!_.has(this.problem, 'solution')) { return null; }

    const stack = [this.problem.solution];
    const texts = [];
    let currentNode;

    while (stack.length) {
      currentNode = stack.pop();
      const { text, children } = this.#extractTextAndChildren(currentNode);
      if (text) { texts.push(text); }
      stack.push(...children);
    }

    return texts.reverse().join('\n ');
  }

  getFeedback(xmlElement) {
    return _.has(xmlElement, 'correcthint') ? _.get(xmlElement, 'correcthint.#text') : '';
  }

  getProblemType() {
    const problemKeys = Object.keys(this.problem);
    const problemTypeKeys = problemKeys.filter(key => Object.values(ProblemTypeKeys).indexOf(key) !== -1);
    if (problemTypeKeys.length === 0) {
      // a blank problem is a problem which contains only `<problem></problem>` as it's olx.
      // blank problems are not given types, so that a type may be selected.
      if (problemKeys.length === 1 && problemKeys[0] === '#text' && this.problem[problemKeys[0]] === '') {
        return null;
      }
      // if we have no matching problem type, the problem is advanced.
      return ProblemTypeKeys.ADVANCED;
    }
    // make sure compound problems are treated as advanced
    // TODO: Find a way to add answers using additional_answers v/s numericalresponse
    if ((problemTypeKeys.length > 1)
      || (problemTypeKeys[0] !== ProblemTypeKeys.NUMERIC // multiple numeric problems are really just multiple answers
        && _.isArray(this.problem[problemTypeKeys[0]])
        && this.problem[problemTypeKeys[0]].length > 1)) {
      return ProblemTypeKeys.ADVANCED;
    }
    const problemType = problemTypeKeys[0];
    return problemType;
  }

  getParsedOLXData() {
    if (_.isEmpty(this.problem)) {
      return {};
    }
    let answersObject = {};
    let additionalAttributes = {};
    let groupFeedbackList = [];
    const problemType = this.getProblemType();
    const hints = this.getHints();
    const question = this.parseQuestions(problemType);
    const solutionExplanation = this.getSolutionExplanation();

    switch (problemType) {
      case ProblemTypeKeys.DROPDOWN:
        answersObject = this.parseMultipleChoiceAnswers(ProblemTypeKeys.DROPDOWN, 'optioninput', 'option');
        break;
      case ProblemTypeKeys.TEXTINPUT:
        answersObject = this.parseStringResponse();
        break;
      case ProblemTypeKeys.NUMERIC:
        answersObject = this.parseNumericResponse();
        break;
      case ProblemTypeKeys.MULTISELECT:
        answersObject = this.parseMultipleChoiceAnswers(ProblemTypeKeys.MULTISELECT, 'checkboxgroup', 'choice');
        break;
      case ProblemTypeKeys.SINGLESELECT:
        answersObject = this.parseMultipleChoiceAnswers(ProblemTypeKeys.SINGLESELECT, 'choicegroup', 'choice');
        break;
      case ProblemTypeKeys.ADVANCED:
        return {
          problemType,
          settings: {},
        };
      default:
        // if problem is unset, return null
        return {};
    }

    if (_.has(answersObject, 'additionalStringAttributes')) {
      additionalAttributes = { ...answersObject.additionalStringAttributes };
    }

    if (_.has(answersObject, 'groupFeedbackList')) {
      groupFeedbackList = answersObject.groupFeedbackList;
    }
    const { answers } = answersObject;
    const settings = { hints };
    if (solutionExplanation) { settings.solutionExplanation = solutionExplanation; }

    return {
      question,
      settings,
      answers,
      problemType,
      additionalAttributes,
      groupFeedbackList,
    };
  }
}
