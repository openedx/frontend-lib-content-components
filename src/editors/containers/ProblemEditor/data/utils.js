/* eslint-disable import/prefer-default-export */
import { replaceRelativeImageUrlsByAbsolute } from '../../../data/services/cms/utils';

export const preprocessProblemData = (problemData, state) => {
  if (problemData?.settings?.hints) {
    problemData.settings.hints = problemData.settings.hints.map(
      hint => ({ ...hint, value: replaceRelativeImageUrlsByAbsolute(hint.value, state.app.lmsEndpointUrl) }),
    );
  }

  if (problemData?.answers) {
    problemData.answers = problemData.answers.map(
      answer => ({ ...answer, title: replaceRelativeImageUrlsByAbsolute(answer.title, state.app.lmsEndpointUrl) }),
    );
  }
};
