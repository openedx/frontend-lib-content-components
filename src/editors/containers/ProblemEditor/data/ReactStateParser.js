import _ from "lodash";
import { ProblemTypeKeys } from "../../../data/constants/problem";
import { indexToLetterMap } from "./MarkDownParser";
import { popuplateItem } from "./SettingsParser";


export class ReactStateParser{
    constructor(problemState){
        this.problemState = problemState;
    }

    parseQuestion() {
        let question = this.problemState.question;
        let paresedQuestion = this.splitLines(question);
        let parsedString = '';
        for (const line in paresedQuestion) {
            let replacedString = paresedQuestion[line].replace("<p>", ">>").replace("</p>", "<<");
            parsedString += `${replacedString}\n`
        }
        return parsedString;
    }

    parseHints() {
        let hintString = '';
        let hitnsList = this.problemState.settings.hints;
        for (const hint in hitnsList){
            hintString += `||${hitnsList[hint].value}||\n`;
        }
        return hintString
    }

    parseDropdown(){
        let answerString = '';
        let answers = this.problemState.answers;
        for (const answer in answers){
            let answerOptionString = '';
            let answerObject = answers[answer];
            let answerContent = answerObject.title;
            let answerFeedback = answerObject?.feedback;
            if (answerObject.correct) {
                answerOptionString += ` (${answerContent})`
            } else {
                answerOptionString += ` ${answerContent}`
            }
            if (answerFeedback !== undefined && answerFeedback.trim() !== ''){
                answerOptionString += ` {{${answerFeedback}}}`
            }
            answerString += `${answerOptionString}\n`
        }
        answerString = `[[\n${answerString}]]\n`;
        return answerString;
    }

    parseCheckBox(){
        let anwerString = '';
        let answers = this.problemState.answers;
        for (const answer in answers){
            let answerObject = answers[answer];
            let answerContent = answerObject.title;
            let answerSelectedFeedback = answerObject?.selectedFeedback;
            let answerUnselectedFeedback = answerObject?.unselectedFeedback;
            let isSelectedFeedbackPresent = (answerSelectedFeedback.trim() !== '' && answerSelectedFeedback !== undefined);
            let isUnselectedFeedbackPresent = (answerUnselectedFeedback.trim() !== '' && answerUnselectedFeedback !== undefined);
            if (answerObject.correct) {
                anwerString += `[x] ${answerContent}`
            } else {
                anwerString += `[ ] ${answerContent}`
            }
            if (isSelectedFeedbackPresent && isUnselectedFeedbackPresent){
                anwerString += `{{selected: ${answerSelectedFeedback}},{unselected: ${answerUnselectedFeedback}}}`
            } else if (isSelectedFeedbackPresent) {
                anwerString += `{{selected: ${answerSelectedFeedback}}}`
            } else if (isUnselectedFeedbackPresent) {
                anwerString += `{{unselected: ${answerUnselectedFeedback}}}`
            }
            anwerString += "\n";
        }
        return anwerString;
    }


    parseRadioButton(){
        let answerString = '';
        let answers = this.problemState.answers;
        for (const answer in answers){
            let answerOptionString = '';
            let answerObject = answers[answer];
            let answerContent = answerObject.title;
            let answerFeedback = answerObject?.feedback;
            if (answerObject.correct) {
                answerOptionString += `(x) ${answerContent}`;
            } else {
                answerOptionString += `( ) ${answerContent}`;
            }
            if (answerFeedback.trim() !== '' && answerFeedback !== undefined){
                answerOptionString += ` {{${answerFeedback}}}`;
            }
            answerString += `${answerOptionString}\n`;
        }
        return answerString;
    }

    parseInputAnswers(){
        let answerString = '';
        let answers = this.problemState.answers;
        let correctAnswer = false;
        for (const answer in answers){
            let answerObject = answers[answer];
            let answerContent = answerObject.title;
            let answerFeedback = answerObject?.feedback;
            let answerOptionString = ''
            if (answerObject.correct && !correctAnswer) {
                answerOptionString += `=${answerContent}`;
                correctAnswer = true;
            } else if (answerObject.correct && correctAnswer) {
                answerOptionString += `or=${answerContent}`;
            } else if (!answerObject.correct){
                answerOptionString += `not=${answerContent}`;
            }
            if (answerFeedback.trim() !== '' && answerFeedback !== undefined){
                answerOptionString += ` {{${answerFeedback}}}`;
            }
            answerString += `${answerOptionString}\n`;
        }
        return answerString;
    }

    parseGroupFeedback() {
        let feedbackString = "";
        const groupFeedbackArray = this.problemState.groupFeedbackList;
        if (!_.isEmpty(groupFeedbackArray)){
            for (const groupFeeback in groupFeedbackArray) {
                let answers = groupFeedbackArray[groupFeeback].answers;
                let options = `(( ${answers.join(" ")} ))`;
                feedbackString += `{{ ${options} ${groupFeedbackArray[groupFeeback].feedback} }}\n`;
            }
        }
        return feedbackString
    }

    getMarkdown() {
        let answers = ''
        switch (this.problemState.problemType) {
            case ProblemTypeKeys.DROPDOWN:
                answers = this.parseDropdown();
                break;
            case ProblemTypeKeys.MULTISELECT:
                answers = this.parseCheckBox();
                break;
            case ProblemTypeKeys.SINGLESELECT:
                answers = this.parseRadioButton();
                break;
            case ProblemTypeKeys.NUMERIC:
            case ProblemTypeKeys.TEXTINPUT:
                answers = this.parseInputAnswers();
                break;
            default:
                break;
        }
        let markdown = `${this.parseQuestion()}${answers}${this.parseHints()}${this.parseGroupFeedback()}`;
        return markdown;
    }

    splitLines(inputString){
        return inputString.trim().split("\n");
    }

    getSettings() {
        let settings = {}
        const stateSettings = this.problemState.settings

        settings = popuplateItem(settings, 'matLabApiKey', 'matlab_api_key', stateSettings)
        settings = popuplateItem(settings, 'number', 'max_attempts', stateSettings.scoring.attempts);
        settings = popuplateItem(settings, 'weight', 'weight', stateSettings.scoring);
        settings = popuplateItem(settings, 'randomization', 'rerandomize', stateSettings);
        settings = popuplateItem(settings, 'on', 'showanswer', stateSettings.showAnswer);
        settings = popuplateItem(settings, 'afterAttempts', 'attempts_before_showanswer_button', stateSettings.showAnswer);
        settings = popuplateItem(settings, 'showResetButton', 'show_reset_button', stateSettings);
        settings = popuplateItem(settings, 'timeBetween', 'submission_wait_seconds', stateSettings);

        return settings;
    }

    getMetadata() {
        markdown = this.getMarkdown();
        settings = this.getSettings();
        return { ...settings, markdown }

    }
}
