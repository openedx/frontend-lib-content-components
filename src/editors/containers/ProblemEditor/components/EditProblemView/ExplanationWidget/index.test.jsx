/* eslint-disable react/prop-types */
import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import { formatMessage } from '../../../../../../testUtils';
import { selectors } from '../../../../../data/redux';
import { ExplanationWidget, mapStateToProps } from '.';

jest.mock('../../../../../data/redux', () => ({
  __esModule: true,
  default: jest.fn(),
  selectors: {
    problem: {
      settings: jest.fn(state => ({ question: state })),
    },
  },
  thunkActions: {
    video: {
      importTranscript: jest.fn(),
    },
  },
}));

jest.mock('../../../../../sharedComponents/TinyMceWidget/hooks', () => ({
  prepareEditorRef: jest.fn(() => ({
    refReady: true,
    setEditorRef: jest.fn().mockName('prepareEditorRef.setEditorRef'),
  })),
}));

describe('SolutionWidget', () => {
  const props = {
    settings: { solutionExplanation: 'This is my question' },
    // injected
    intl: { formatMessage },
  };
  describe('render', () => {
    test('snapshot: renders correct default', () => {
      expect(shallow(<ExplanationWidget {...props} />).snapshot).toMatchSnapshot();
    });
  });
  describe('mapStateToProps', () => {
    const testState = { A: 'pple', B: 'anana', C: 'ucumber' };
    test('question from problem.question', () => {
      expect(mapStateToProps(testState).settings).toEqual(selectors.problem.settings(testState));
    });
  });
});
