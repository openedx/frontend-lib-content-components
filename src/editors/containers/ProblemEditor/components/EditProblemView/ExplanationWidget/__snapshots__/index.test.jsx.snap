// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SolutionWidget render snapshot: renders correct default 1`] = `
<div
  className="tinyMceWidget mt-4 text-primary-500"
>
  <div
    className="h4 mb-3"
  >
    <FormattedMessage
      defaultMessage="Explanation"
      description="Explanation Title"
      id="authoring.problemEditor.explanationwidget.explanationWidgetTitle"
    />
  </div>
  <div
    className="small mb-3"
  >
    <FormattedMessage
      defaultMessage="Provide an explanation for the correct answer"
      description="Description of the solution widget"
      id="authoring.problemEditor.solutionwidget.solutionDescriptionText"
    />
  </div>
  <[object Object]
    editorContentHtml="This is my solution"
    editorType="solution"
    id="solution"
    minHeight={150}
    placeholder="Enter your explanation"
    setEditorRef={[MockFunction prepareEditorRef.setEditorRef]}
  />
</div>
`;
